cpp/src/dictionary/remove_keys.cu

```diff
@@ -0,0 +1,175 @@
+#include <cudf/detail/copy_if.cuh>
+#include <cudf/detail/gather.hpp>
+#include <cudf/detail/valid_if.cuh>
+#include <cudf/detail/search.hpp>
+#include <cudf/table/table.hpp>
+#include <cudf/table/table_view.hpp>
+
+#include <thrust/sequence.h>
+#include <thrust/scatter.h>
+#include <thrust/transform.h>
+
+namespace cudf
+{
+namespace dictionary
+{
+namespace detail
+{
+namespace
+{
+
+/**
+ * @brief Return a new dictionary by removing identified keys from the provided dictionary.
+ *
+ * This is a common utility for `remove_keys` and `remove_unused_keys` detail functions.
+ * It will create a new dictionary with the remaining keys and create new indices values
+ * to go with these new keys.
+ *
+ * @tparam KeysKeeper Function bool(size_type) that takes keys position index
+ *                    and returns true if that key is to be used in the output dictionary.
+ * @param dictionary_column The column to use for creating the new dictionary.
+ * @param keys_to_keep_fn Called to determine which keys in `dictionary_column` to keep.
+ * @param mr Resource for creating output columns.
+ * @param stream CUDA Stream for kernel calls.
+ */
+template<typename KeysKeeper>
+std::unique_ptr<column> remove_keys_fn( dictionary_column_view const& dictionary_column,
+                                        KeysKeeper keys_to_keep_fn,
+                                        rmm::mr::device_memory_resource* mr = rmm::mr::get_default_resource(),
+                                        cudaStream_t stream = 0)
+{
+    // create keys positions column to identify original key positions after removing they keys
+    auto keys_view = dictionary_column.keys();
+    auto execpol = rmm::exec_policy(stream);
+    rmm::device_vector<int32_t> keys_positions(keys_view.size()); // needed for remapping indices
+    thrust::sequence( execpol->on(stream), keys_positions.begin(), keys_positions.end() );
+    column_view keys_positions_view( data_type{INT32}, keys_view.size(), keys_positions.data().get() );
+
+    // copy the non-removed keys ( keys_to_keep_fn(idx)==true )
+    std::unique_ptr<column> keys_column;
+    rmm::device_vector<int32_t> map_indices(keys_view.size(),-1); // init -1 to identify new nulls
+    {
+        auto table_keys = experimental::detail::copy_if( table_view{{keys_view, keys_positions_view}},
+                                                         keys_to_keep_fn, mr, stream )->release();
+        keys_column = std::move(table_keys.front());
+        keys_positions_view = table_keys[1]->view();
+        // build indices mapper
+        // Example scatter([0,1,2][0,2,4][-1,-1,-1,-1,-1]) => [0,-1,1,-1,2]
+        thrust::scatter( execpol->on(stream), thrust::make_counting_iterator<int32_t>(0),
+                         thrust::make_counting_iterator<int32_t>(keys_positions_view.size()),
+                         keys_positions_view.begin<int32_t>(), map_indices.begin() );
+    } // frees up the temporary table_keys objects
```

Comment:
Is it possible to initialize `keys_column` with an immediately invoked lambda? Also removes the "floating braces" is so. Something like:
```cpp
    rmm::device_vector<int32_t> map_indices(keys_view.size(),-1); // init -1 to identify new nulls
    std::unique_ptr<column> keys_column = [&] {
        auto table_keys = experimental::detail::copy_if( table_view{{keys_view, keys_positions_view}},
                                                         keys_to_keep_fn, mr, stream )->release();
        std::unique_ptr<column> keys_column = std::move(table_keys.front());
        keys_positions_view = table_keys[1]->view();
        // build indices mapper
        // Example scatter([0,1,2][0,2,4][-1,-1,-1,-1,-1]) => [0,-1,1,-1,2]
        thrust::scatter( execpol->on(stream), thrust::make_counting_iterator<int32_t>(0),
                         thrust::make_counting_iterator<int32_t>(keys_positions_view.size()),
                         keys_positions_view.begin<int32_t>(), map_indices.begin() );
      return std::move(keys_column);
    } ();
```